File: bindings/typescript/src/assemblers/token.ts
import { FragmentId, TokenPrimitiveFragment } from '@vizstack/schema';
import { FragmentAssembler } from '../fragment-assembler';

type ColorType = TokenPrimitiveFragment['contents']['color'];

class TokenPrimitiveFragmentAssembler extends FragmentAssembler {
    private _text: string;
    private _color?: ColorType;

    constructor(
        text: string,
        color?: ColorType,
    ) {
        super();
        this._text = text;
        this._color = color;
    }

    public assemble(getId: (obj: any, name: string) => FragmentId): [TokenPrimitiveFragment, any[]] {
        return [
            {
                type: 'TokenPrimitive',
                contents: {
                    text: this._text,
                    color: this._color,
                },
                meta: this._meta,
            },
            [],
        ];
    }
}

export function Token(
    text: string,
    color?: ColorType,
) {
    return new TokenPrimitiveFragmentAssembler(text, color);
}

export interface Token extends ReturnType<typeof Token> {}

File: bindings/python/vizstack/assemblers/token.py
from vizstack.fragment_assembler import FragmentAssembler
from typing import Optional, Tuple, Dict, List, Any
from typing_extensions import Literal
from vizstack.schema import JsonType, View, Fragment


Color = Literal['gray', 'brown', 'purple', 'blue', 'green', 'yellow', 'orange', 'red', 'pink', None]


class Token(FragmentAssembler):
    """
    A View which renders a boxed piece of text on a colored background.
    """

    def __init__(self,
                 text: str,
                 color: Color = None) -> None:
        """
        Args:
            text: Text which should be rendered.
            color: Name of color for background.
        """
        super(Token, self).__init__()
        self._text: str = text
        self._color: Color = color

    def assemble(self, get_id) -> Tuple[Fragment, List[Any]]:
        return {
            'type': 'TokenPrimitive',
            'contents': {
                'text': self._text,
                'color': self._color,
            },
            'meta': self._meta,
        }, []
